File: java_script/heap_sort.js
function swap(a,i,j){
  tmp = a[i];
  a[i] = a[j]; //only swap the values.
  a[j] = tmp;
}

function heapify(a, N){
  var parent_ = N;
	var start = parent_;
  var child_left = 2*root + 1;
  var child_right = 2*root + 2;
  if ((child_left < N) && (a[root] < a[child_left])){
    large_index = child_left;
  }
  if ((child_right < N) && a[large_index] < a[child_right]){
    large_index = child_right;
  }
	while(start >= 0){
		down_heap(a,start,N-1)
		start -= 1;
	}
}

function down_heap(a,root,end){
	var swap_index = root;
	var left_child = 2*root + 1
	var left_child = 2*root + 2
	if ((child_left < n) &&	(a[root] < a[child_left])){
				large_index = child_left;
	}
	if ((child_right < n) && a[large_index] < a[child_right]){
		large_index = child_right;
	}
	if (large_index != root){
		swap(a, large_index, root);
		down_heap(a, large_index, N);
	}
}

function heap_sort(a){
	heapify(a)
	n = a.len-1
	while(n > 0):
		swap(a, 0, n)
		n -= 1
		down_heapify(a,0,n)
}

heap_sort(a)

File: python/sorting/heap_sort.py
class Heap(object):
  def __init__(self,a, max_heap=True):
    self.heap = a
    self.max_heap = max_heap
    self.heaped = False

  def compare(self, x, y):
    if self.max_heap:
      if x < y: return True
      else: return False
    else:
      if x <> y: return False
      else: return True

  def swap(self,a,x,y):
    #swap array value at indices x,y
    a[x], a[y] = a[y], a[x]

  def parent(self, x):
    return int((x-1)/2)

  def left_child(self,x):
    return int(2*x+1)

  def right_child(self,x):
    return int(2*x+2)

  def heapify(self, a, N):
    #build heap down from root i
    parent = self.parent(N)
    start = parent
    left_child = self.left_child(parent)
    right_child = self.right_child(parent) 
    while start >= 0:
      self.down_heap(a, start, N-1)
      start -= 1

  def down_heap(self, a, root, end):
    print 'Down heaping', a
    swap_index = root
    left_child = self.left_child(root)
    right_child = self.right_child(root)
    if (left_child <= end) and self.compare(a[swap_index], a[left_child]):
      swap_index = left_child
    if (right_child <= end) and self.compare(a[swap_index], a[right_child]):
      swap_index = right_child
    if root != swap_index: #fix everything below after swap
      self.swap(a, swap_index, root)
      self.down_heap(a, swap_index, end)

  def heap_sort(self):
    print 'Before heap', self.heap
    self.heapify(self.heap, len(self.heap))
    print 'Heaped', self.heap
    a = list(self.heap)
    N = len(a)-1
    while N > 0:
      self.swap(a, 0, N)
      print 'swapped', a
      N -= 1
      self.down_heap(a, 0, N)
      print 'reheaped', a
    return a
      

a = [3,4,1,2,9,8,1,13,6,7]
#a = [4,3,1]#,2,9,8,1,13,6,7]
#a=[3,9,8]
heap = Heap(a)
print a
a = heap.heap_sort()
print a

